### PR TITLE
processing data in models

### PR DESCRIPTION
**Pull request by Zeph180**
Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
Set up associations between models.
Remember that author_id column in posts table should be the foreign_key for the users table.
### Create custom methods

1. Users
- A method that returns the 3 most recent posts for a given user.

2.  Posts
- A method that updates the posts counter for a user.
- A method which returns the 5 most recent comments for a given post.
3. Comments
- A method that updates the comments counter for a post.
4. Likes
- A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id', dependent: :destroy
+  has_many :comments, foreign_key: 'author_id'
+
```
Comment (mikethreels) [inline app/models/user.rb]: - [ ] as you can see in the below image from the ERD diagram there should be a has_many relationship for the likes as well in the user class.
afbeelding

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class Post < ApplicationRecord
+  belongs_to :user
+  has_many :comments, foreign_key: 'author_id'
+  has_many :likes, class_name: 'Like', foreign_key: 'post_id', dependent: :destroy
```
Comment (mikethreels) [inline app/models/post.rb]: - [ ] In the user model you named the foreign key Author_id therefore you should mention this in the belongs to part of the relation as well like below:
```
belongs_to :author, class_name: 'User'
```

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class Post < ApplicationRecord
+  belongs_to :user
+  has_many :comments, foreign_key: 'author_id'
```
Comment (mikethreels) [inline app/models/post.rb]: - [ ] it looks like you might have copied and pasted this line form the user model, which is no problem, however, make sure to adjust it to the post model as well. please remove the foreign_key: 'author_id' in this line

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
Comment (mikethreels) [inline app/models/like.rb]: - [ ] In the user model you named the foreign key Author_id therefore you should mention this in the belongs to part of the relation as well like below:
```
belongs_to :author, class_name: 'User'

File: app/models/comment.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
```
Comment (mikethreels) [inline app/models/comment.rb]: - [ ] good job adding this line, however, you can remove the foreign_key part this is only set in the has_many part of the relationship.

File: app/models/comment.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class Comment < ApplicationRecord
+  belongs_to :user
```
Comment (mikethreels) [inline app/models/comment.rb]: - [ ] having added the belongs to part with author you can remove this belongs_to relation.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Zeph, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- linters are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/like.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
```
Comment (danmainah) [inline app/models/like.rb]: - [ ] [optiional] @Zeph180 great job implementing all changes requested, however, I would recommend removing the `foreign key` part, this is usually set in a has-many part of the relationship, you don't need it here.

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id', dependent: :destroy
+  has_many :comments, foreign_key: 'author_id'
+  has_many :likes, class_name: 'Like', foreign_key: 'author_id', dependent: :destroy
```
Comment (danmainah) [inline app/models/user.rb:4]: - [ ]  [optiional] @Zeph180 Great job implementing all changes requested, however, I would recommend removing the class_name part, this is usually set in a belongs_to part of the relationship, you don't need it here.

**Review by danmainah — APPROVED** [review]
Hi @Zeph180 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 

## Highlights
✔️ Github flow was correctly used
✔️  All association setup correctly

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**